pkg/ccl/backupccl/manifest_handling.go

```diff
@@ -1234,8 +1235,10 @@ func checkForPreviousBackup(
 }
 
 // tempCheckpointFileNameForJob returns temporary filename for backup manifest checkpoint.
+// It writes to its own temp-progress directory so to not interfere
+// with how normal checkpoint files are stored and read.
 func tempCheckpointFileNameForJob(jobID jobspb.JobID) string {
-	return fmt.Sprintf("%s-%d", backupManifestCheckpointName, jobID)
+	return fmt.Sprintf("TEMP-%s-%d", backupManifestCheckpointName, jobID)
```

Comment:
Hmm okay, just trying to wrap my head around this. I think I mostly agree (just trying to convince myself) with your logic.

I understand why the non-mixed version cluster would violate write-once semantics, and the fix seems pretty straight forward. Since it's 22.1 only we can just write a normal checkpoint right? 
Or more specifically, instead of calling `writeBackupManifestCheckpoint`, we just call `writeBackupManifest` and pass it in NewTimestampedCheckpointFileName() as the filename. Which would functionally write the same thing just no TEMP prefix. 
And then readManifestOnResume would just treat it as though there was already a checkpoint written which shouldn't be a problem? Since it doesn't really care if it was a temp checkpoint or a regular one? If that logic follows then I agree that it would work in this case.

Planning on 21.2 node logic seems sound.

Planning on a 22.1 node seems pretty similar to the logic for non-mixed to me. We don't actually care if it a temp checkpoint or actual checkpoint, we just want the latest one so it's okay to do away with the TEMP prefix. 
I guess the only thing I'm a little iffy about is the `non-timestamp suffixed since we are in a mixed version state` part. I understand why you'd have to do this for the base directory, cause that's where 21.2 nodes will look. But does this apply to the progress directory? 
Because otherwise this would mess with listing (since BACKUP-CHECKPOINT will always be at the top). I don't see why we couldn't version the checkpoint we write into the progress directory unless I'm missing something.

Assuming that my reasoning falls in line with yours, I agree that we can do away with the TEMP prefix + jobID suffix all together.

Reply:
>Since it doesn't really care if it was a temp checkpoint or a regular one?

exactly, yup. My slight preference here would be to have `writeBackupManifestCheckpoint` do all the timestamping and writing of the checkpoint and checksum, and `writeBackupManifest` only be responsible for writing the manifest. The special casing to check if the filename is a checkpoint or manifest makes me uneasy, so even if it means some duplication of code I think I'm okay with that.

> But does this apply to the progress directory?

You're right, so you're saying if a 22.1 node does the planning, and we're in a mixed version state, we should write a non-ts manifest checkpoint to the base directory and this way 21.2 and 22.1 nodes that may execute the backup job will see this?

Reply:
> writeBackupManifestCheckpoint do all the timestamping and writing of the checkpoint and checksum,

Sure, good with me. I also agree the checking of filename to see what to do with it seems bad.

> we should write a non-ts manifest checkpoint to the base directory

Yeah so I was thinking if it's a mixed version, we just call the `writeNewCheckpointFile` WITH the timestamp suffixed. Which should do two things, write a non-ts checkpoint in the base directory, write a ts checkpoint in the progress directory.

1. If a 22.1 node executes the job, it will `readManifestOnResume` which first checks the progress directory, which will find our ts checkpoint properly since it looks in progress first.

2. If a 21.2 node executes the job, it will also call `readManifestOnResume`, but instead only check the base directory, which will find our non-ts checkpoint that we wrote.

Like I guess technically the write into the progress directory is unneeded, but as the logic stands right now, there is no further implementation needed to write to both. Whereas only writing to base would require some sort of flag or separate function to tell it it's not actually a regular checkpoint but rather a temp checkpoint. And I feel like that would be making the code a lot messier? Unless theres a reason I'm missing as to why we can't write to progress.